azurelinuxagent/ga/guestagent.py

```diff
@@ -0,0 +1,314 @@
+        logger.verbose(msg)
+        add_event(
+            AGENT_NAME,
+            version=self.version,
+            op=WALAEventOperation.Install,
+            is_success=True,
+            message=msg)
+
+    def _ensure_loaded(self):
+        self._load_manifest()
+        self._load_error()
+
+    def _download(self):
+        try:
+            self._protocol.client.download_zip_package("agent package", self.pkg.uris, self.get_agent_pkg_path(), self.get_agent_dir(), use_verify_header=self._is_fast_track_goal_state)
+        except Exception as exception:
+            msg = "Unable to download Agent {0}: {1}".format(self.name, ustr(exception))
+            add_event(
+                AGENT_NAME,
+                op=WALAEventOperation.Download,
+                version=CURRENT_VERSION,
+                is_success=False,
+                message=msg)
+            raise UpdateError(msg)
+
+    def _load_error(self):
+        try:
+            self.error = GuestAgentError(self.get_agent_error_file())
+            self.error.load()
+            logger.verbose(u"Agent {0} error state: {1}", self.name, ustr(self.error))
+        except Exception as e:
+            logger.warn(u"Agent {0} failed loading error state: {1}", self.name, ustr(e))
+
+    def _load_manifest(self):
+        path = self.get_agent_manifest_path()
+        if not os.path.isfile(path):
+            msg = u"Agent {0} is missing the {1} file".format(self.name, AGENT_MANIFEST_FILE)
+            raise UpdateError(msg)
+
+        with open(path, "r") as manifest_file:
+            try:
+                manifests = json.load(manifest_file)
+            except Exception as e:
+                msg = u"Agent {0} has a malformed {1} ({2})".format(self.name, AGENT_MANIFEST_FILE, ustr(e))
+                raise UpdateError(msg)
+            if type(manifests) is list:
+                if len(manifests) <= 0:
+                    msg = u"Agent {0} has an empty {1}".format(self.name, AGENT_MANIFEST_FILE)
+                    raise UpdateError(msg)
+                manifest = manifests[0]
+            else:
+                manifest = manifests
+
+        try:
+            self.manifest = HandlerManifest(manifest)  # pylint: disable=W0201
+            if len(self.manifest.get_enable_command()) <= 0:
+                raise Exception(u"Manifest is missing the enable command")
+        except Exception as e:
+            msg = u"Agent {0} has an illegal {1}: {2}".format(
+                self.name,
+                AGENT_MANIFEST_FILE,
+                ustr(e))
+            raise UpdateError(msg)
+
+        logger.verbose(
+            u"Agent {0} loaded manifest from {1}",
+            self.name,
+            self.get_agent_manifest_path())
+        logger.verbose(u"Successfully loaded Agent {0} {1}: {2}",
+                       self.name,
+                       AGENT_MANIFEST_FILE,
+                       ustr(self.manifest.data))
+        return
+
+
+class GuestAgentError(object):
```

Comment:
GuestAgentError used in GuestAgent and no other places have a reference, so I kept in same module.